network/crl_test.go

```diff
@@ -0,0 +1,350 @@
+ZG9uMRgwFgYDVQQKDA9HbG9iYWwgU2VjdXJpdHkxCzAJBgNVBAsMAklUMRwwGgYD
+VQQDDBNUZXN0IENBIGNlcnRpZmljYXRlMCAXDTIwMTAyNjIzMjcwNVoYDzIwNzAx
+MDE0MjMyNzA1WjCBjDELMAkGA1UEBhMCR0IxDzANBgNVBAgMBkxvbmRvbjEPMA0G
+A1UEBwwGTG9uZG9uMRgwFgYDVQQKDA9HbG9iYWwgU2VjdXJpdHkxCzAJBgNVBAsM
+AklUMTQwMgYDVQQDDCtUZXN0IGxlYWYgY2VydGlmaWNhdGUgKGFjcmEtd3JpdGVy
+LXJldm9rZWQpMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAwYeEHPae
+eZmQTn0XVbk7A4cokbF5aNg2mqNfgLDPBScEW6iN0Jm0/61JLDU67TGG0Lkan3qZ
+51p+V8CW+geG8gWcHRKHOHUsaZQhkO94niFD2rTJclbJfBpza/sIdji11lmH5oOi
+uzvBbX+4BtopierZdEXysa3WoFpM7rIxryOYeXegDjiDFqY7RmZvs9cD9hB+SDaX
+W7whTRh8jFWOmGVEAq4YbiqbkfaFEX+8YrFoAjl4zHFlXfE3K7Wvx/WxGYuS50+4
+4Wofkd6afnVggqWli3gLn/iYLjbog94vYJ0CSBf13NfYn8frMa01O9GGHsuShjNJ
+MDt+/ZfPoHPVmQIDAQABo4GUMIGRMAkGA1UdEwQCMAAwCwYDVR0PBAQDAgXgMBQG
+A1UdEQQNMAuCCWxvY2FsaG9zdDAxBggrBgEFBQcBAQQlMCMwIQYIKwYBBQUHMAGG
+FWh0dHA6Ly8xMjcuMC4wLjE6ODg4ODAuBgNVHR8EJzAlMCOgIaAfhh1odHRwOi8v
+MTI3LjAuMC4xOjg4ODkvY3JsLnBlbTANBgkqhkiG9w0BAQsFAAOCAQEAH4yh4kUT
+wAxm7cKNu73hOGwz9HFswp44jpAl4q9/kJHkA/ycITFbHEPKsruWfqLk7vxp7kgW
+8Fv+qMyP5b0njDtAUWQuV8gn+DRQzPun8scG0kGJQb24ZQ4nWEPFm42H/roYv47B
+7y7cYmnycP9c4oCaQMdJ29w4bdaC9SHZdQW0W57UkmelVjFU00odxia5O7VhKRkP
+FC+mY9/+xkj2aDvunWLHwE4wXSzrFpOkGqNPpRU47HNO3K3giRTZXtKCap61J55B
+W8Q1chrNalbgBuYfsbtVlCEoVuzPmkjGzJIHASk8NNo3jQ7EgQJUbff8d1MZwa2f
+k5rkNiHTxPfguQ==
+-----END CERTIFICATE-----
+`)
+}
+
+func getTestCRL() []byte {
+	// x509.ParseCRL() that is called inside CRL verifier is very fragile
+	// and will return error because of \n on start, that's why this func
+	// looks a little bit different from getTest*Cert()
+
+	// tests/ssl/crl.pem
+	return []byte(`-----BEGIN X509 CRL-----
+MIIB4zCBzDANBgkqhkiG9w0BAQsFADB0MQswCQYDVQQGEwJHQjEPMA0GA1UECAwG
+TG9uZG9uMQ8wDQYDVQQHDAZMb25kb24xGDAWBgNVBAoMD0dsb2JhbCBTZWN1cml0
+eTELMAkGA1UECwwCSVQxHDAaBgNVBAMME1Rlc3QgQ0EgY2VydGlmaWNhdGUXDTIw
+MTAyNjIzMjcwNloXDTIwMTEyNTIzMjcwNlowJzAlAhQH6gdHXIXsVUmzREIYCF6n
+ajAdThcNMjAxMDI2MjMyNzA2WjANBgkqhkiG9w0BAQsFAAOCAQEAKKJ56Ba7bOqE
+R8LFhfSwqqzGxVdPfDOu3d14cpRC3PJSLcyfLh+YA8Rkveed7Ee1RGaGjXrXTiZm
+p44CQcslUFoFGdONVTzTAamCphOiLh/LZB65d2dWo1WwOfsLwwaE+KTk+99+Ug/j
+EgiyuOwoyFqONlaDj4Z2PUC/mKmPuI1QJ9WB43oL8cDxPuOtJp0z2NEz0ZDl1pu6
+ghaEHc7lPDBmCOOGYneqEtfAi/wzMzKEdeXxV9s3MBWsXKfdrwsQc2pjLuH+K/wv
+aY3BPTUA7sZLJGFvsE2fh9yi1TsyKs4hsboS6FldZdXBnD/0ZqUnFy/KQ4t3LwH5
+L6zfD+wegQ==
+-----END X509 CRL-----`)
+}
+
+// Convert PEM-encoded certificate into DER and parse it into x509.Certificate
+func pemToX509Cert(data []byte) (*x509.Certificate, error) {
+	certDERBlock, _ := pem.Decode(data)
+	if certDERBlock == nil {
+		return nil, errors.New("Failed to decode PEM")
+	}
+
+	if certDERBlock.Type != "CERTIFICATE" {
+		return nil, errors.New("Decoded DER block was not a certificate")
+	}
+
+	cert, err := x509.ParseCertificate(certDERBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+func TestDefaultCRLClientHTTP(t *testing.T) {
+	crl := getTestCRL()
+	addr := "127.0.0.1:45318"
```

Comment:
I guess it would be (technically) better to use `:0` here to let the OS choose a random unused port for the server.

First you manually create a listening socket:
```go
listener, err := net.Listen("tcp", ":0")
```
then you pass it to the server:
```go
go httpServer.Serve(listener)
```
and connect later to the listener's address:
```go
uri := fmt.Sprintf("http://%s/test_crl.pem", listener.Addr().String())
```

This approach also has a hidden benefit of creating the listening socket beforehand, as opposed to using `ListenAndServe` in a goroutine. That way the client can start connecting right away, as the listening socket is already there, so you don't need a `time.Sleep()` call to avoid errors.

Reply:
Wow, didn't know you can let OS decide the port to listen on